File: lib/widgets/notification_popup.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';

import '../notification_service.dart';

/// Every notification an app is currently showing. The arrow on an app tile -
/// or a swipe right across it - opens this instead of guessing which one you
/// meant; tapping a row fires that notification's own action and swiping it
/// away clears it, the way pulling down the shade would.
Future<void> showAppNotifications(
  BuildContext context, {
  required String appName,
  required List<AppNotification> notifications,
  Uint8List? icon,
}) {
  // A copy: the list handed in belongs to NotificationService, and swiping a
  // row here only clears it until the next snapshot arrives from the shade.
  final items = List.of(notifications);
  final isDarkMode = Theme.of(context).brightness == Brightness.dark;
  final textColor = isDarkMode ? Colors.white : Colors.black;
  final subtleColor = isDarkMode ? Colors.white70 : Colors.black54;

  return showDialog(
    context: context,
    builder: (dialogContext) {
      return Dialog(
        backgroundColor:
            isDarkMode ? const Color(0xFF252525) : Colors.white.withAlpha(242),
        insetPadding: const EdgeInsets.symmetric(horizontal: 24, vertical: 48),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(20)),
        // Keeps a swipe's red panel inside the rounded corners.
        clipBehavior: Clip.antiAlias,
        child: StatefulBuilder(
          builder: (context, setDialogState) {
            return ConstrainedBox(
              constraints: BoxConstraints(
                maxHeight: MediaQuery.of(context).size.height * 0.7,
              ),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Padding(
                    padding: const EdgeInsets.fromLTRB(20, 20, 20, 16),
                    child: Row(
                      children: [
                        if (icon != null)
                          ClipRRect(
                            borderRadius: BorderRadius.circular(10),
                            child: Image.memory(icon,
                                width: 40, height: 40, fit: BoxFit.cover),
                          )
                        else
                          Container(
                            width: 40,
                            height: 40,
                            decoration: BoxDecoration(
                              color: isDarkMode
                                  ? const Color(0xFF424242)
                                  : const Color(0xFFE0E0E0),
                              borderRadius: BorderRadius.circular(10),
                            ),
                            child: const Icon(Icons.apps, size: 22),
                          ),
                        const SizedBox(width: 15),
                        Expanded(
                          child: Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(
                                appName,
                                maxLines: 1,
                                overflow: TextOverflow.ellipsis,
                                style: TextStyle(
                                  color: textColor,
                                  fontSize: 16,
                                  fontWeight: FontWeight.bold,
                                ),
                              ),
                              Text(
                                '${items.length} notification'
                                '${items.length == 1 ? '' : 's'}',
                                style:
                                    TextStyle(color: subtleColor, fontSize: 14),
                              ),
                            ],
                          ),
                        ),
                      ],
                    ),
                  ),
                  const Divider(height: 1),
                  Flexible(
                    child: ListView.separated(
                      shrinkWrap: true,
                      padding: EdgeInsets.zero,
                      itemCount: items.length,
                      separatorBuilder: (_, __) => const Divider(height: 1),
                      itemBuilder: (context, index) {
                        final notification = items[index];
                        return Dismissible(
                          key: ValueKey(notification.key),
                          background: Container(
                            color: Colors.red.withAlpha(180),
                            padding: const EdgeInsets.symmetric(horizontal: 24),
                            child: const Row(
                              mainAxisAlignment: MainAxisAlignment.spaceBetween,
                              children: [
                                Icon(Icons.clear, color: Colors.white),
                                Icon(Icons.clear, color: Colors.white),
                              ],
                            ),
                          ),
                          onDismissed: (_) {
                            NotificationService.dismiss(notification.key);
                            setDialogState(() => items.removeAt(index));
                            if (items.isEmpty) Navigator.pop(dialogContext);
                          },
                          child: ListTile(
                            contentPadding: const EdgeInsets.symmetric(
                                horizontal: 20, vertical: 8),
                            title: Row(
                              crossAxisAlignment: CrossAxisAlignment.baseline,
                              textBaseline: TextBaseline.alphabetic,
                              children: [
                                Expanded(
                                  child: Text(
                                    notification.title.isEmpty
                                        ? appName
                                        : notification.title,
                                    maxLines: 1,
                                    overflow: TextOverflow.ellipsis,
                                    style: TextStyle(
                                      color: textColor,
                                      fontSize: 15,
                                      fontWeight: FontWeight.w600,
                                    ),
                                  ),
                                ),
                                Text(
                                  shortAgo(notification.postTime),
                                  style: TextStyle(
                                      color: subtleColor, fontSize: 12),
                                ),
                              ],
                            ),
                            subtitle: notification.text.isEmpty
                                ? null
                                : Padding(
                                    padding: const EdgeInsets.only(top: 4),
                                    child: Text(
                                      notification.text,
                                      style: TextStyle(
                                          color: subtleColor, fontSize: 14),
                                    ),
                                  ),
                            onTap: () {
                              Navigator.pop(dialogContext);
                              NotificationService.open(notification.key);
                            },
                          ),
                        );
                      },
                    ),
                  ),
                ],
              ),
            );
          },
        ),
      );
    },
  );
}

File: app/src/main/kotlin/com/kayfahaarukku/fuselauncher/ui/NotificationPopup.kt
package com.kayfahaarukku.fuselauncher.ui

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Apps
import androidx.compose.material.icons.filled.Clear
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.SwipeToDismissBox
import androidx.compose.material3.SwipeToDismissBoxValue
import androidx.compose.material3.Text
import androidx.compose.material3.rememberSwipeToDismissBoxState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import com.kayfahaarukku.fuselauncher.notifications.AppNotification
import com.kayfahaarukku.fuselauncher.notifications.shortAgo

private val TextColor = Color.White
private val SubtleColor = Color.White.copy(alpha = 0.7f)
private val DividerColor = Color.White.copy(alpha = 0.12f)

/**
 * Every notification an app is currently showing.
 *
 * Tapping a row fires that notification's own action and swiping it away
 * clears it, the way pulling down the shade would. Colours are the Flutter
 * build's fixed greys rather than the Material scheme, which tints surfaces
 * towards the primary colour and reads as the wrong grey next to the original.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NotificationPopup(
    appName: String,
    icon: ImageBitmap?,
    notifications: List<AppNotification>,
    onOpen: (AppNotification) -> Unit,
    onDismissNotification: (AppNotification) -> Unit,
    onDismiss: () -> Unit,
) {
    // A copy: the caller's list is replaced wholesale on the next snapshot from
    // the shade, so a row swiped here should stay gone until then.
    var items by remember(notifications) { mutableStateOf(notifications) }
    val maxHeight = LocalConfiguration.current.screenHeightDp.dp * 0.7f

    LaunchedEffect(items) {
        if (items.isEmpty()) onDismiss()
    }

    Dialog(onDismissRequest = onDismiss) {
        Column(
            Modifier
                .padding(horizontal = 8.dp)
                .clip(RoundedCornerShape(20.dp))
                .background(PopupSurface)
                .heightIn(max = maxHeight),
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = 20.dp, top = 20.dp, end = 20.dp, bottom = 16.dp),
            ) {
                if (icon != null) {
                    Image(
                        bitmap = icon,
                        contentDescription = null,
                        modifier = Modifier.size(40.dp).clip(RoundedCornerShape(10.dp)),
                    )
                } else {
                    Box(
                        Modifier
                            .size(40.dp)
                            .clip(RoundedCornerShape(10.dp))
                            .background(Color(0xFF424242)),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(Icons.Filled.Apps, null, tint = TextColor, modifier = Modifier.size(22.dp))
                    }
                }
                Spacer(Modifier.width(15.dp))
                Column(Modifier.weight(1f)) {
                    Text(
                        appName,
                        color = TextColor,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Bold,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                    Text(
                        "${items.size} notification${if (items.size == 1) "" else "s"}",
                        color = SubtleColor,
                        fontSize = 14.sp,
                    )
                }
            }

            PopupDivider()

            LazyColumn {
                items(items, key = { it.key }) { notification ->
                    val dismissState = rememberSwipeToDismissBoxState()

                    LaunchedEffect(dismissState.currentValue) {
                        if (dismissState.currentValue != SwipeToDismissBoxValue.Settled) {
                            onDismissNotification(notification)
                            items = items - notification
                        }
                    }

                    SwipeToDismissBox(
                        state = dismissState,
                        backgroundContent = {
                            Row(
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically,
                                modifier = Modifier
                                    .fillMaxSize()
                                    .background(BadgeRed.copy(alpha = 0.7f))
                                    .padding(horizontal = 24.dp),
                            ) {
                                Icon(Icons.Filled.Clear, "Dismiss", tint = Color.White)
                                Icon(Icons.Filled.Clear, "Dismiss", tint = Color.White)
                            }
                        },
                    ) {
                        NotificationRow(notification, appName) { onOpen(notification) }
                    }
                    PopupDivider()
                }
            }
        }
    }
}

@Composable
private fun PopupDivider() = Box(
    Modifier.fillMaxWidth().height(1.dp).background(DividerColor)
)

@Composable
private fun NotificationRow(
    notification: AppNotification,
    appName: String,
    onClick: () -> Unit,
) = Column(
    Modifier
        .fillMaxWidth()
        .background(PopupSurface)
        .clickable(onClick = onClick)
        .padding(horizontal = 20.dp, vertical = 8.dp),
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Text(
            notification.title.ifEmpty { appName },
            color = TextColor,
            fontSize = 15.sp,
            fontWeight = FontWeight.SemiBold,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f),
        )
        Spacer(Modifier.width(8.dp))
        Text(shortAgo(notification.postTime), color = SubtleColor, fontSize = 12.sp)
    }
    if (notification.text.isNotEmpty()) {
        Text(
            notification.text,
            color = SubtleColor,
            fontSize = 14.sp,
            modifier = Modifier.padding(top = 4.dp),
        )
    }
}
